src/query/api/v1/handler/database/create.go

```diff
@@ -375,82 +378,185 @@ func defaultedNamespaceAddRequest(
+	default:
+		// Use the maximum block size if we don't find a way to
+		// recommended one based on request parameters
+		max := recommendedBlockSizesByRetentionAsc[len(recommendedBlockSizesByRetentionAsc)-1]
+		blockSize = max.blockSize
+		for _, elem := range recommendedBlockSizesByRetentionAsc {
+			if retentionPeriod <= elem.forRetentionLessThanOrEqual {
+				blockSize = elem.blockSize
+				break
 			}
+		}
+	}
 
-			if blockSize < minRecommendCalculateBlockSize {
-				blockSize = minRecommendCalculateBlockSize
-			} else if blockSize > maxRecommendCalculateBlockSize {
-				blockSize = maxRecommendCalculateBlockSize
-			}
+	retentionOpts = retentionOpts.SetBlockSize(blockSize)
 
-		default:
-			// Use the maximum block size if we don't find a way to
-			// recommended one based on request parameters
-			max := recommendedBlockSizesByRetentionAsc[len(recommendedBlockSizesByRetentionAsc)-1]
-			blockSize = max.blockSize
-			for _, elem := range recommendedBlockSizesByRetentionAsc {
-				if retentionPeriod <= elem.forRetentionLessThanOrEqual {
-					blockSize = elem.blockSize
-					break
-				}
-			}
+	indexOpts := opts.IndexOptions().
+		SetEnabled(true).
+		SetBlockSize(blockSize)
+
+	opts = opts.SetRetentionOptions(retentionOpts).
+		SetIndexOptions(indexOpts)
+
+	// Resolution does not apply to unaggregated namespaces so set to 0.
+	attrs, err := dbnamespace.NewAggregationAttributes(0, dbnamespace.Unaggregated)
+	if err != nil {
+		return nil, err
+	}
+
+	opts = opts.SetAggregationOptions(dbnamespace.NewAggregationOptions().
+		SetAggregations([]dbnamespace.Aggregation{
+			dbnamespace.NewAggregation(attrs, dbnamespace.NewDownsampleOptions(true)),
+		}))
+
+	return &admin.NamespaceAddRequest{
+		Name:    r.NamespaceName,
+		Options: dbnamespace.OptionsToProto(opts),
+	}, nil
+}
+
+func defaultedAggregatedNamespaceAddRequest(
+	r *admin.DatabaseCreateRequest,
+) (*admin.NamespaceAddRequest, error) {
+	agg := r.AggregationNamespace
+	if agg == nil {
+		return nil, nil
+	}
+
+	if agg.Name == "" {
+		return nil, errors.New("name required when aggregationNamespace is set")
+	}
+
+	if agg.Resolution == "" {
+		return nil, errors.New("resolution required when aggregationNamespace is set")
+	}
+
+	if agg.RetentionTime == "" {
+		return nil, errors.New("retention_time required when aggregationNamespace is set")
+	}
 
+	opts := dbnamespace.NewOptions().
+		SetRepairEnabled(false)
+
+	retentionOpts := opts.RetentionOptions()
+	value, err := time.ParseDuration(agg.RetentionTime)
+	if err != nil {
+		return nil, fmt.Errorf("invalid retention time: %v", err)
+	}
+	retentionOpts = retentionOpts.SetRetentionPeriod(value)
+
+	// Use the maximum block size if we don't find a way to
+	// recommended one based on request parameters
+	retentionPeriod := retentionOpts.RetentionPeriod()
+	max := recommendedBlockSizesByRetentionAsc[len(recommendedBlockSizesByRetentionAsc)-1]
+	blockSize := max.blockSize
+	for _, elem := range recommendedBlockSizesByRetentionAsc {
```

Comment:
I think some of this blockSize, retention, and resolution setting code is in both agg and unagg code paths so maybe we can share in private funcs